### ISSUE TITLE
Controller decorator not working in config admin yaml

### ISSUE THREAD

**Issue by Flowster** [PrestaShop/PrestaShop#24662]
#### Describe the bug

When in a module you declare a controller service decorator under `config/admin/services.yml` an exception is raised.
When you declare this decorator under `config/services.yml` it's working fine.

#### Expected behavior

We should be able to declare the decorator in admin or front context.

#### Steps to Reproduce

1. From this PR #18811 download the [testdecorationcontrollers.zip](https://github.com/PrestaShop/PrestaShop/files/4645455/testdecorationcontrollers.zip) module zip file.
2. Move `config/services.yml` to `config/admin/services.yml`.
3. Install the module and you get an error.

![image](https://user-images.githubusercontent.com/2620053/119648519-d9128580-be21-11eb-8e5f-e19050c8478f.png)

#### Additional information

* PrestaShop version: 1.7.7.4
* PHP version: 7.3

#### Related documentation

* [Services in Legacy environment](https://devdocs.prestashop.com/1.7/modules/concepts/services/#services-in-legacy-environment)
* [Decorate the controller](https://devdocs.prestashop.com/1.7/modules/concepts/controllers/admin-controllers/override-decorate-controller/#decorate-the-controller)

![image](https://user-images.githubusercontent.com/2620053/119693867-ff014f80-be4c-11eb-9297-fd0201c2c253.png)

**Comment by hibatallahAouadni** [comment]
Hello @Flowster 

Could you please explain to me why you need to move config/services.yml to config/admin/services.yml :sweat_smile:
Please check and feedback.

Thanks!

**Comment by Flowster** [comment]
I might have mistaken the purpose of those config/admin/services.yml files...

I first thought that you should choose between having a config/services.yml file and config/admin-front/services.yml files.

But I now understand that those are two distinct with different purpose systems. The first one for full Symfony services and the second one only for Legacy ones.

I wrongly thought you had to choose between those two concepts but we shall use both together.

You can then close this issue if that's the case.

**Comment by matks** [comment]
Hi @Flowster I actually believe your issue is right: AFAIK config/admin/services.yml files have the same purpose for different environments.

If I am not wrong admin/services.yml and front/services.yml have been added later to allow the use of services more widely than initial services.yml . @jolelievre can you confirm? I see no reason to use simple services.yml now (but we keep it probably for backward compatibility)

So @Flowster your bug report seems valid IMO.

I'll be honest: decoration is a powerful but complex feature of Symfony, I already encountered [one bug related to it](https://github.com/symfony/symfony/issues/36567) so it's totally possible that something related to decorators work for config/services.yml and does not work for config/admin/services.yml 😞 

> When in a module you declare a controller service decorator under config/admin/services.yml an exception is raised.
When you declare this decorator under config/services.yml it's working fine.

It looks like a bug behavior for me, something that needs to be fixed.

**Comment by jolelievre** [comment]
@matks @Flowster Ok I understand the problem now!

So indeed as the documentation explains there are three different services config files, but `config/admin/services.yml` and `config/front/services.yml` are handled by our legacy container which is a light Container used in legacy front/admin controllers. They are both initialized in those places:

https://github.com/PrestaShop/PrestaShop/blob/948ecce8461176c507512c977633731bc3550a24/classes/controller/AdminController.php#L4724
https://github.com/PrestaShop/PrestaShop/blob/756b3a11f2d982fa3d9215566b86a93be1708950/classes/controller/FrontController.php#L1994

This light container is not built by the Symfony framework, so it doesn't contain all the Symfony components and, as a result, not the Symfony controllers either. It is still useful if you need to define some services that are accessible in front. And as you explained you can create a `config/common/services.yml` that is included in the three states:
- legacy front
- legacy admin
- Symfony admin

So all this explains why you can't decorate a controller in `config/admin/services.yml` and have to do it in `config/services.yml` I hope this clarifies things, this multiple container context is not easy to understand at first, I tried to make it as clear as possible in the documentation but I guess it can still be improved.

I'd be interested if you have ideas on how it could be improved? (I'm not objective anymore and fail to see the missing info)

## News Flash in 1.7.8

So after this long explanation, here's a great news: starting 1.7.8 there is no more legacy admin container, even in legacy admin controller we now only use the Symfony container:

https://github.com/PrestaShop/PrestaShop/blob/316fd48b693510ce44ddb5fb5841233e1bc2885e/classes/controller/AdminController.php#L4762

This change was introduced by this PR if you want more details: https://github.com/PrestaShop/PrestaShop/pull/20413

So tldr; @Flowster what you are trying to do should work with the 1.7.8. Can I ask you a favor? Could you try decorating the controller from `config/amin/services.yml` using the 1.7.8.x branch just to confirm that it now works?

**Comment by hibatallahAouadni** [comment]
Hello @Flowster 

Since we had no news from you for more than 20 days, I'll close this ticket to avoid cluttering up the backlog.
Please note that you can always create a new one if further information pops up.

Thanks!
